contracts/smart-wallet/utils/BaseAccountFactory.sol

```diff
@@ -0,0 +1,102 @@
+// $$$$$$\   $$$$$$$\  $$\  $$$$$$\   $$$$$$$ |$$\  $$\  $$\  $$$$$$\  $$$$$$$\
+// \_$$  _|  $$  __$$\ $$ |$$  __$$\ $$  __$$ |$$ | $$ | $$ |$$  __$$\ $$  __$$\
+//   $$ |    $$ |  $$ |$$ |$$ |  \__|$$ /  $$ |$$ | $$ | $$ |$$$$$$$$ |$$ |  $$ |
+//   $$ |$$\ $$ |  $$ |$$ |$$ |      $$ |  $$ |$$ | $$ | $$ |$$   ____|$$ |  $$ |
+//   \$$$$  |$$ |  $$ |$$ |$$ |      \$$$$$$$ |\$$$$$\$$$$  |\$$$$$$$\ $$$$$$$  |
+//    \____/ \__|  \__|\__|\__|       \_______| \_____\____/  \_______|\_______/
+
+abstract contract BaseAccountFactory is IAccountFactory, Multicall {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /*///////////////////////////////////////////////////////////////
+                                State
+    //////////////////////////////////////////////////////////////*/
+
+    BaseAccount internal immutable _accountImplementation;
+
+    mapping(address => EnumerableSet.AddressSet) internal accountsOfSigner;
+    mapping(address => EnumerableSet.AddressSet) internal signersOfAccount;
+
+    /*///////////////////////////////////////////////////////////////
+                            Constructor
+    //////////////////////////////////////////////////////////////*/
+
+    constructor(BaseAccount accountImpl) {
+        _accountImplementation = accountImpl;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        External functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Deploys a new Account for admin.
+    function createAccount(address _admin, bytes calldata _data) external virtual returns (address);
+
+    /// @notice Callback function for an Account to register its signers.
+    function addSigner(address _signer) external {
+        address account = msg.sender;
+
+        bool isAlreadyAccount = accountsOfSigner[_signer].add(account);
+        bool isAlreadySigner = signersOfAccount[account].add(_signer);
+
+        if (!isAlreadyAccount || !isAlreadySigner) {
+            revert("AccountFactory: signer already added");
+        }
+
+        emit SignerAdded(account, _signer);
+    }
+
+    /// @notice Callback function for an Account to un-register its signers.
+    function removeSigner(address _signer) external {
+        address account = msg.sender;
+
+        bool isAccount = accountsOfSigner[_signer].remove(account);
+        bool isSigner = signersOfAccount[account].remove(_signer);
+
+        if (!isAccount || !isSigner) {
+            revert("AccountFactory: signer not found");
+        }
+
+        emit SignerRemoved(account, _signer);
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            View functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Returns the implementation of the Account.
+    function accountImplementation() external view override returns (address) {
```

Comment:
why is this external? would be cleaner to use this rather than exposing `_accountImplementation` no?